src/client/Microsoft.Identity.Client/Credential/ManagedIdentityCredentialResponse.cs

```diff
@@ -0,0 +1,175 @@
+            }
+            else
+            {
+                if (credentialResponse.Credential.IsNullOrEmpty())
+                {
+                    errorMessages.Add("Credential is missing or empty. ");
+                }
+
+                if (credentialResponse.RegionalTokenUrl.IsNullOrEmpty())
+                {
+                    errorMessages.Add("RegionalTokenUrl is missing or empty. ");
+                }
+
+                if (credentialResponse.ClientId.IsNullOrEmpty())
+                {
+                    errorMessages.Add("ClientId is missing or empty. ");
+                }
+
+                if (credentialResponse.TenantId.IsNullOrEmpty())
+                {
+                    errorMessages.Add("TenantId is missing or empty. ");
+                }
+            }
+
+            // Check if any error messages were added
+            if (errorMessages.Any())
+            {
+                // Log an error message indicating the missing or insufficient fields
+                _requestContext.Logger.Error("[Managed Identity] " + string.Join(" ", errorMessages) +
+                    " and/or insufficient for authentication.");
+
+                // Throw an exception indicating that the CredentialResponse is invalid
+                MsalException exception = MsalServiceExceptionFactory.CreateManagedIdentityException(
+                    MsalError.ManagedIdentityRequestFailed,
+                    MsalErrorMessage.ManagedIdentityInvalidResponse,
+                    null,
+                    ManagedIdentitySource.Credential,
+                    null);
+
+                throw exception;
+            }
+        }
+
+        /// <summary>
+        /// Fetches a new managed identity credential from the IMDS endpoint.
+        /// </summary>
+        /// <param name="httpManager"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="MsalManagedIdentityException"></exception>
+        private async Task<CredentialResponse> FetchFromServiceAsync(
+            IHttpManager httpManager,
+            CancellationToken cancellationToken)
+        {
+            _requestContext.Logger.Info("[Managed Identity] Fetching new managed identity credential from IMDS endpoint.");
+
+            OAuth2Client client = CreateClientRequest(httpManager);
+
+            CredentialResponse credentialResponse = await client
+                .GetCredentialResponseAsync(_uri, _requestContext)
+                .ConfigureAwait(false);
+
+            return credentialResponse;
+        }
+
+        /// <summary>
+        /// Creates an OAuth2 client request for fetching the managed identity credential.
+        /// </summary>
+        /// <param name="httpManager"></param>
+        /// <returns></returns>
+        private OAuth2Client CreateClientRequest(IHttpManager httpManager)
+        {
+            var client = new OAuth2Client(_requestContext.Logger, httpManager, null);
```

Comment:
Hmm, OAuth2Client is general purpose, but for contacting the token endpoint we use `TokenClient` which does throttling and HTTP telemetry.